File: chunk.py
#!/usr/bin/python


def rule(x, z):
    def base(x, z):
        a = (x/3 + z/3) % 2 == 0
        b = x % 3 == 0 and z % 3 == 0
        return a and b
    n = -1
    if base(x, z): n = 0
    if base(x + 1, z): n = 1
    if base(x, z + 1): n = 2
    if base(x + 1, z + 1): n = 3

    b = ((x+2)/6 + (z+2)/6) % 4
    if b == n: return '*'
    if n == -1: return ' '
    return ' '

out = ''
for x in xrange(-20, 25):
    for z in xrange(-80, 80):
        out += rule(x, z)
    out += '\n'

print(out)

File: chunk.c
#include <stdio.h>
#define false 0
#define true 1

char base(int x, int z) {
    char a = (x/4 + z/4) % 3 == 0;
    char b = (x % 4 == 0) && (z % 4 == 0);
    return a && b;
}

char rule(int x, int z) {
    if (x < 0) x = -x;
    if (z < 0) z = -z;
    char n = -1;
    int N = ((x-1)/8 + (z+1)/8) % 4;
    if (base(x, z)) n = 0;
    else if (base(x + 1, z)) n = 1;
    else if (base(x, z + 1)) n = 2;
    else if (base(x + 1, z + 1)) n = 3;
    else {
        //return "012345"[N];
        return ' ';
    }
    //return "012345678"[n];
    if (N == n) return '*';
    return ' ';
}

void main() {
    for (int x = 0; x <= 45; x++) {
        for (int z = 0; z <= 80*2; z++) {
            putchar(rule(x, z));
        }
        putchar('\n');
    }
}
